File: packages/api-client/src/poe2.ts
import {isPoe2Build, isPoe2Job, isPoe2Comparison, type Poe2Build, type Poe2Job, type Poe2Changes, type Poe2Comparison} from '@wow-mini/domain'
import type {ClientAuthContext} from './auth-context'
import type {ApiResult, ApiTransport, RequestData} from './transport'
import {apiV2Path} from './api-v2-prefix'
import {isPoe2Import, type Poe2Import, type Poe2CreateImportRequest} from '@wow-mini/domain'
import {isPoe2Tree, type Poe2Tree} from '@wow-mini/domain'

export interface Poe2Client {
  getTree(buildId: string, auth: ClientAuthContext, jobId?: string): Promise<ApiResult<Poe2Tree | null>>
  createImport(body: Poe2CreateImportRequest, auth: ClientAuthContext): Promise<ApiResult<Poe2Import | null>>
  getImport(importId: string, auth: ClientAuthContext): Promise<ApiResult<Poe2Import | null>>
  supplyImportSource(importId: string, body: {source: string; idempotencyKey: string}, auth: ClientAuthContext): Promise<ApiResult<Poe2Import | null>>
  retryImport(importId: string, body: {idempotencyKey: string}, auth: ClientAuthContext): Promise<ApiResult<Poe2Import | null>>
  cancelImport(importId: string, auth: ClientAuthContext): Promise<ApiResult<Poe2Import | null>>
  listBuilds(auth: ClientAuthContext): Promise<ApiResult<{items: Poe2Build[]}>>
  deleteBuild(id: string, auth: ClientAuthContext): Promise<ApiResult<{deleted: boolean}>>
  importBuild(body: {source: string; title?: string; gameVersion?: string; league?: string}, auth: ClientAuthContext): Promise<ApiResult<Poe2Build | null>>
  getBuild(id: string, auth: ClientAuthContext): Promise<ApiResult<Poe2Build | null>>
  listJobs(buildId: string, auth: ClientAuthContext): Promise<ApiResult<{items: Poe2Job[]}>>
  calculate(body: {buildId: string; changes: Poe2Changes; idempotencyKey: string}, auth: ClientAuthContext): Promise<ApiResult<Poe2Job | null>>
  getJob(id: string, auth: ClientAuthContext): Promise<ApiResult<Poe2Job | null>>
  compare(jobIds: [string, string], auth: ClientAuthContext): Promise<ApiResult<Poe2Comparison | null>>
  exportBuild(id: string, auth: ClientAuthContext): Promise<ApiResult<{exportCode: string}>>
  craftingLink(itemText: string, auth: ClientAuthContext): Promise<ApiResult<{url: string}>>
}

const record = (v: unknown): v is Record<string, unknown> => typeof v === 'object' && v !== null && !Array.isArray(v)
const id = (value: string): string => {
  if (!value || value.length > 128) throw new Error('无效记录标识')
  return encodeURIComponent(value)
}
export function createPoe2Client(transport: ApiTransport): Poe2Client {
  const request = <T>(path: string, auth: ClientAuthContext, fallback: () => T,
    validate: (v: unknown) => boolean, body?: RequestData, timeoutMs?: number): Promise<ApiResult<T>> => transport.request(apiV2Path('/poe2' + path), {
      auth, responseMode: 'structured-problem', fallback, validate,
      ...(timeoutMs === undefined ? {} : {timeoutMs}),
      ...(body === undefined ? {} : {method: 'POST' as const, data: body}),
    })
  return {
    getTree: (buildId, auth, jobId) => request('/builds/' + id(buildId) + '/tree' + (jobId ? '?jobId=' + id(jobId) : ''), auth, () => null, isPoe2Tree, undefined, 65000),
    createImport: (body, auth) => request('/imports', auth, () => null, isPoe2Import, {...body}),
    getImport: (importId, auth) => request('/imports/' + id(importId), auth, () => null, isPoe2Import),
    supplyImportSource: (importId, body, auth) => request('/imports/' + id(importId) + '/source', auth, () => null, isPoe2Import, body),
    retryImport: (importId, body, auth) => request('/imports/' + id(importId) + '/retry', auth, () => null, isPoe2Import, body),
    cancelImport: (importId, auth) => request('/imports/' + id(importId) + '/cancel', auth, () => null, isPoe2Import, {}),
    listBuilds: auth => request('/builds', auth, () => ({items: []}), v => record(v) && Array.isArray(v['items']) && v['items'].every(isPoe2Build)),
    deleteBuild: (buildId, auth) => request('/builds/' + id(buildId) + '/delete', auth, () => ({deleted: false}), v => record(v) && v['deleted'] === true, {}),
    importBuild: (body, auth) => request('/builds', auth, () => null, isPoe2Build, body, 65000),
    getBuild: (buildId, auth) => request('/builds/' + id(buildId), auth, () => null, isPoe2Build),
    listJobs: (buildId, auth) => request('/jobs?buildId=' + id(buildId), auth, () => ({items: []}), v => record(v) && Array.isArray(v['items']) && v['items'].every(isPoe2Job)),
    calculate: (body, auth) => request('/jobs', auth, () => null, isPoe2Job, body as unknown as RequestData),
    getJob: (jobId, auth) => request('/jobs/' + id(jobId), auth, () => null, isPoe2Job),
    compare: (jobIds, auth) => request('/compare', auth, () => null, isPoe2Comparison, {jobIds}),
    exportBuild: (buildId, auth) => request('/builds/' + id(buildId) + '/export', auth, () => ({exportCode: ''}), v => record(v) && typeof v['exportCode'] === 'string' && Boolean(v['exportCode'])),
    craftingLink: (itemText, auth) => request('/crafting/import-link', auth, () => ({url: ''}), v => record(v) && typeof v['url'] === 'string' && v['url'].startsWith('https://beta.craftofexile.com/?'), {itemText}),
  }
}

File: server/app/api/routes/poe2.py
from typing import Any
from uuid import UUID
from fastapi import APIRouter, Depends, Header, Query, Request, Response
from pydantic import BaseModel, ConfigDict, Field

from server.app.api.dependencies import require_mutating_principal, require_principal, poe2_application
from server.app.api.errors import ApiProblem
from server.app.identity.domain import Principal
from server.app.poe2.application import Poe2Application, Poe2Error, build_packet, job_packet
from server.app.poe2.tools import Poe2ToolUnauthorized

router=APIRouter(prefix='/api/v2/poe2',tags=['poe2'])
internal_router=APIRouter()

class Strict(BaseModel):model_config=ConfigDict(extra='forbid')
class ImportBody(Strict):
    source:str
    title:str='Imported build'
    gameVersion:str=''
    league:str=''
class JobBody(Strict):
    buildId:UUID
    changes:dict[str,Any]|None=None
    idempotencyKey:str=Field(min_length=1,max_length=128)
class CompareBody(Strict):jobIds:list[UUID]
class CraftBody(Strict):itemText:str
class ToolBody(Strict):operation:str; arguments:dict[str,Any]

def _call(fn):
    try:return fn()
    except Poe2Error as e:raise ApiProblem(status_code=e.status,code=e.code,message=e.code) from None
    except ValueError as e:
        code=str(e) if str(e).startswith('POE2_') else 'POE2_REQUEST_INVALID'
        raise ApiProblem(status_code=422,code=code,message=code) from None

@router.post('/builds',status_code=201)
def import_build(body:ImportBody,p:Principal=Depends(require_mutating_principal),a:Poe2Application=Depends(poe2_application)):
    return _call(lambda:build_packet(a.import_build(p,body.source,title=body.title,game_version=body.gameVersion,league=body.league),include_source=True))
@router.get('/builds')
def list_builds(limit:int=Query(50,ge=1,le=100),p:Principal=Depends(require_principal),a:Poe2Application=Depends(poe2_application)):
    return {'items':[build_packet(x) for x in a.list_builds(p,limit)]}
@router.get('/builds/{build_id}')
def get_build(build_id:UUID,p:Principal=Depends(require_principal),a:Poe2Application=Depends(poe2_application)):
    return _call(lambda:build_packet(a.read_build(p,build_id),include_source=True))
@router.post('/builds/{build_id}/delete')
def delete_build(build_id:UUID,p:Principal=Depends(require_mutating_principal),a:Poe2Application=Depends(poe2_application)):
    return _call(lambda:a.delete_build(p,build_id))
@router.get('/builds/{build_id}/export')
def export(build_id:UUID,p:Principal=Depends(require_principal),a:Poe2Application=Depends(poe2_application)):
    return _call(lambda:a.export(p,build_id))
@router.get('/builds/{build_id}/tree')
def tree(build_id:UUID,response:Response,jobId:UUID|None=None,p:Principal=Depends(require_principal),a:Poe2Application=Depends(poe2_application)):
    response.headers['Cache-Control']='private, no-store'
    return _call(lambda:a.read_tree(p,build_id,jobId))
@router.post('/jobs',status_code=202)
def submit(body:JobBody,p:Principal=Depends(require_mutating_principal),a:Poe2Application=Depends(poe2_application)):
    return _call(lambda:job_packet(a.submit_job(p,body.buildId,body.changes,body.idempotencyKey)))
@router.get('/jobs')
def list_jobs(buildId:UUID|None=None,limit:int=Query(50,ge=1,le=100),p:Principal=Depends(require_principal),a:Poe2Application=Depends(poe2_application)):
    return _call(lambda:{'items':[job_packet(x) for x in a.list_jobs(p,buildId,limit)]})
@router.get('/jobs/{job_id}')
def get_job(job_id:UUID,p:Principal=Depends(require_principal),a:Poe2Application=Depends(poe2_application)):
    return _call(lambda:job_packet(a.read_job(p,job_id)))
@router.post('/compare')
def compare(body:CompareBody,p:Principal=Depends(require_mutating_principal),a:Poe2Application=Depends(poe2_application)):
    return _call(lambda:a.compare(p,body.jobIds))
@router.post('/crafting/import-link')
def craft(body:CraftBody,p:Principal=Depends(require_mutating_principal),a:Poe2Application=Depends(poe2_application)):
    return _call(lambda:a.crafting_link(body.itemText))

@internal_router.post('/api/v2/internal/chickenbro/poe2-tool')
def tool(body:ToolBody,request:Request,x_chickenbro_poe2_gateway:str=Header(default='')):
    try:return request.app.state.chickenbro_poe2_gateway.execute(x_chickenbro_poe2_gateway,body.operation,body.arguments)
    except Poe2ToolUnauthorized:
        raise ApiProblem(status_code=401,code='CHAT_EXECUTION_INACTIVE',message='capability unavailable') from None
    except (Poe2Error,ValueError,KeyError,TypeError) as e:
        code=e.code if isinstance(e,Poe2Error) else 'CHAT_TOOL_UNAVAILABLE'
        raise ApiProblem(status_code=422,code=code,message=code) from None


from server.app.api.dependencies import poe2_import_application
from server.app.poe2.imports.application import ImportApplication, packet_json

class CharacterImportBody(Strict):
    provider:str
    url:str=Field(max_length=4096)
    idempotencyKey:str=Field(min_length=1,max_length=128)
class CharacterSourceBody(Strict):
    source:str=Field(max_length=2000000)
    idempotencyKey:str=Field(min_length=1,max_length=128)
class CharacterRetryBody(Strict):
    idempotencyKey:str=Field(min_length=1,max_length=128)

@router.post('/imports',status_code=202)
def create_character_import(body:CharacterImportBody,p:Principal=Depends(require_mutating_principal),a:ImportApplication=Depends(poe2_import_application)):
    return _call(lambda:packet_json(a.create(p,body.provider,body.url,body.idempotencyKey)))

@router.get('/imports/{import_id}')
def read_character_import(import_id:UUID,p:Principal=Depends(require_principal),a:ImportApplication=Depends(poe2_import_application)):
    return _call(lambda:packet_json(a.read(p,import_id)))

@router.post('/imports/{import_id}/source',status_code=202)
def supply_character_source(import_id:UUID,body:CharacterSourceBody,p:Principal=Depends(require_mutating_principal),a:ImportApplication=Depends(poe2_import_application)):
    return _call(lambda:packet_json(a.supply_source(p,import_id,body.source,body.idempotencyKey)))

@router.post('/imports/{import_id}/retry',status_code=202)
def retry_character_import(import_id:UUID,body:CharacterRetryBody,p:Principal=Depends(require_mutating_principal),a:ImportApplication=Depends(poe2_import_application)):
    return _call(lambda:packet_json(a.retry(p,import_id,body.idempotencyKey)))

@router.post('/imports/{import_id}/cancel')
def cancel_character_import(import_id:UUID,p:Principal=Depends(require_mutating_principal),a:ImportApplication=Depends(poe2_import_application)):
    return _call(lambda:packet_json(a.cancel(p,import_id)))
